# POLI3148 Lecture Notes: Week 4 Lecture 4

This week we covered the basics of data wrangling with Pandas, which focuses on the second part of the whole data science workflow, i.e., data processing. One natural question to start with is: why do we bother doing data wrangling? The answer is that often times we can't have perfect datasets that can be analyzed right away. You might encounter missing values, wrong numbers, different data shapes, or you might want to create additional columns to store some results of your calculations. Let's say you have a bunch of dates and time. Some are written in the format of dd/mm/yyyy hh:mm, some are reversed and separated by dashes like this: yyyy-mm-dd, some use military time but some have that AM/PM suffix. You want to clean them to keep them in a consistent format. Or if you are analyzing the economic performance of several countries, their data are all in different currency, which you have to convert to compare apples to apples. Countries might also have different items in their domestic economic survey. You might need to create a unique indicator of economic performance out of these statistics, or select only countries your research question cares about.

There are of course more things Python can do. We will go through them one by one.

## The case for this lecture: V-Dem

Although we have introduced the dataset in the last lecture, here are some more information about it: The dataset we are using can be found through v-dem.net. It has a lot of inidcators of various regimes from democracy to dictatorship, with socioeconomic statistics collected from other sources. We have touched a bit of it in previous lectures when we used the GDP of Mexico and compulsory voting. More info about this dataset, you can go to their webset to check out. If possibile, you can use the full .csv file. (By the way: What is a CSV file? CSV is Comma Separated Values (they really are) and are commonly used in data science. You can read using almost any text editor. Excel, Numbers, and many other programming languages also accept it as a valid format for data storage.)

This VDem dataset can be used to inspire some ideas for your final project. Some other datasets are also recommended:

Political/ socioeconomic data:
* Quality of Government: https://www.gu.se/en/quality-government/qog-data/data-downloads
* World Bank: https://databank.worldbank.org/databases
* United Nations: https://databank.worldbank.org/databases

Opinion data:
* World Value Survey: https://www.worldvaluessurvey.org/
* Global Barometer Survey: https://www.globalbarometer.net/

A comprehensive list of all political databases:
* https://github.com/erikgahner/PolData



**Table of contents**

1. Setting up project folder

2. Data importing and exporting

3. Data processing
   * Subsetting
   * Transforming
   * Aggregating
   * Renaming
   * Combining

4. Logistics

At the end of today's lecture, you should know how to:

* import: `pd.read_csv`
* overview: `df.shape`, `df.head()`, `df.tail()`, `df.describe()`, `df.columns`, and `df.info`
* subset: `df.loc` and `df.iloc`
* transform
* aggregate `.agg`
* rename  `.rename`
* combine `pd.merge` and `pd.concat`
* export `.to_csv`

...the data you have. We will also talk about some good practices of data management.

## 1. Setting up the project folder

**The platform**

To use examples in this lecture notes, you will need to link your Google Drive to this specific notebook. Click the folder button on the left column and you will see an "sample_data" folder. Click on the Google Drive button above it to mount your Drive. You will be prompted to login and authorize.

**The data**

The dataset we are using can be found in either the Classroom or through v-dem.net. It has a lot of inidcators of various regimes from democracy to dictatorship, with socioeconomic statistics collected from other sources. We have touched a bit of it in previous lectures when we used the GDP of Mexico and compulsory voting. More info about this dataset, you can go to their webset to check out. If possibile, you can use the full .csv file. (By the way: What is a CSV file? CSV is Comma Separated Values (they really are) and are commonly used in data science. You can read using almost any text editor. Excel, Numbers, and many other programming languages also accept it as a valid format for data storage.)

This VDem dataset can be used to inspire some ideas for your final project. Some other datasets are also recommended:

Political/ socioeconomic data:
* Quality of Government: https://www.gu.se/en/quality-government/qog-data/data-downloads
* World Bank: https://databank.worldbank.org/databases
* United Nations: https://databank.worldbank.org/databases

Opinion data:
* World Value Survey: https://www.worldvaluessurvey.org/
* Global Barometer Survey: https://www.globalbarometer.net/

A comprehensive list of all political databases:
* https://github.com/erikgahner/PolData

**The project folder**

Finally, let's set up our project folder. It is important to keep good folder organization for your projects. Dr. Chen mentioned one of the good practices for data scientists when you oragnizing your folder for data: Put documentation, schema, readme file, and other things that can help your collaborator to understand more about the dataset you are using inside the same folder along side with that dataset. For our case of VDem data, there are a couple of these files that are available when you download it from the website. So make sure you oragnize your folder that way.

Another thing I want to mention about project folder oragnization is that you should also create subfolders that are clear enough for anyone that looks at it the first time. You can have your code stored at the root of our project folder, followed by a dedicated folder for your data, another one for data that was processed, another one for all of your plots, maybe also another one if you are doing some kind of write-up, such as manuscripts and posters. For this lecture, you are recommended to have a project folder named "Lecture_4", an ipynb file inside it, a folder named "data" where the VDem data is stored, and a folder named "data_processed" to store any modifications we make to the original VDem dataset.

You are recommended to use a similar way of folder organization for your assignments and final projects. Good oragnization is so important that doing so can really make your life much much easier. Some other rules of thumb are listed below for your reference. Whenever you are thinking "Huh, I'll just leave this thing here and will come back and tidy it up" you should certainly look at these rules and stop doing that.

Generally:
1. Document your data sources and information of retrieval
2. Organize your data files (don’t throw them to random places!)
3. Keep documentation (codebook, user manuals etc.)
4. Always keep the raw data unaltered

For cllaborative project:
1. Work on the same copy of data.
2. If you are doing Google Drive-based collaboration, non-owners should add “shortcuts” of the folder
3. and all members add a shortcut to the folder at the root folder ⇒ Everyone can run everyone else’s code.

**From now on, you should follow these good practcies while coding in classes, and while you do exercises, assignments, and projects. Failure to do so will lead to reduction of points.**

**Pandas library**

A Python library contains a lot of useful functionalities that might not be found in base Python. Developers who work on data sciences come up with these tools for us. Pandas is one of them. It specializes in data wrangling and data cleaning. To get pandas working in your project, you can:


In [ ]:
import pandas as pd

Remember we talked about we do not want to type in all these letters every time we use pandas? That's what the "as pd" does for us. It assigns a shortened name to pandas. We will see more shorthands of Python libraries in the future.

## 2. Data importing and exporting

If you still remember the three big steps when doing data work, the very first one is data imporing. That is to store the data in the memory with a name attached to it so that Python can access it whenever it is needed. To do so, we are learning the very first pandas command as below.

In [ ]:
# Replace this with your path
df = pd.read_csv("/content/drive/MyDrive/POLI3148/Lecture4/data/V-Dem-CY-Full+Others-v14.csv")

<ipython-input-2-5f980c35e97c>:1: DtypeWarning: Columns (364,365,366,399,415,804,836,837,924,1240,1257,1486,3094,3168,3169,3341,3342,3344,3345,3347,3350,3352) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("/content/drive/MyDrive/POLI3148/POLI3148_Lecture4/data/V-Dem-CY-Full+Others-v14.csv")


Remember, the path to your data must be a string, i.e., quoted texts inside that pair of parentheses. The VDem data is now assigned to a name called df. Also, this is just my path inside my folder, yours is almost certainly different from mine. So make sure you change that part before running any following codes.

We talked a bit about CSV before, but it is still worth mentioning again. You can check it out on Wikipedia here: https://en.wikipedia.org/wiki/Comma-separated_values

There are some other similar types of data, such as TSV (Tab Separated Values).

After initially importing the dataset, we might want to give it an overview. What does it look like? There are some useful commands you can use, the first of which is df.shape() and df.columns(), or df.columns.tolist().

In [ ]:
df.shape

(27734, 4607)

This shows that we have 27734 rows and 4507 columns in this dataframe. That is very large!

In [ ]:
df.columns

Index(['country_name', 'country_text_id', 'country_id', 'year',
       'historical_date', 'project', 'historical', 'histname', 'codingstart',
       'codingend',
       ...
       'e_mipopula', 'e_miurbani', 'e_pefeliex', 'e_wb_pop', 'e_pechmor',
       'e_miinteco', 'e_civil_war', 'e_miinterc', 'e_pt_coup',
       'e_pt_coup_attempts'],
      dtype='object', length=4607)

This lists all columns of this dataframe, with some of them omitted in the middle. If you want to see all of them, the following is very useful.

In [ ]:
df.columns.tolist()

['country_name',
 'country_text_id',
 'country_id',
 'year',
 'historical_date',
 'project',
 'historical',
 'histname',
 'codingstart',
 'codingend',
 'codingstart_contemp',
 'codingend_contemp',
 'codingstart_hist',
 'codingend_hist',
 'gapstart1',
 'gapstart2',
 'gapstart3',
 'gapend1',
 'gapend2',
 'gapend3',
 'gap_index',
 'COWcode',
 'v2x_polyarchy',
 'v2x_polyarchy_codelow',
 'v2x_polyarchy_codehigh',
 'v2x_polyarchy_sd',
 'v2x_libdem',
 'v2x_libdem_codelow',
 'v2x_libdem_codehigh',
 'v2x_libdem_sd',
 'v2x_partipdem',
 'v2x_partipdem_codelow',
 'v2x_partipdem_codehigh',
 'v2x_partipdem_sd',
 'v2x_delibdem',
 'v2x_delibdem_codelow',
 'v2x_delibdem_codehigh',
 'v2x_delibdem_sd',
 'v2x_egaldem',
 'v2x_egaldem_codelow',
 'v2x_egaldem_codehigh',
 'v2x_egaldem_sd',
 'v2x_api',
 'v2x_api_codelow',
 'v2x_api_codehigh',
 'v2x_api_sd',
 'v2x_mpi',
 'v2x_mpi_codelow',
 'v2x_mpi_codehigh',
 'v2x_mpi_sd',
 'v2x_freexp_altinf',
 'v2x_freexp_altinf_codelow',
 'v2x_freexp_altinf_codehigh',
 'v2

Note that Python might force you into thinking of the dataset using columns and rows. It is nice to do so, but sometimes we want to be a bit less technical. We usually call the columns "variables", and rows "records", or "observations". Why?

Because a column is often a property or "thing" that can have multiple values across all rows. If we are looking at a dataset that shows the GDP of some countries, we might have a column named "GDP", which can have many possible values. This makes it variable, i.e., something that can vary. Whereas each country in the rows will become an observation of this or a group of properties.

Like the data types we learned in previous lectures, we also care about that data type our dataset have. Run the following code to check it.

In [ ]:
df.dtypes

,0
country_name,object
country_text_id,object
country_id,int64
year,int64
historical_date,object
...,...
e_miinteco,float64
e_civil_war,float64
e_miinterc,float64
e_pt_coup,float64


There are different types of columns. Some are object, some are int64, some are float64. You might be familiar with the last two. They mean that all values inside that column is integer or float, i.e., numbers. But what is object? Object can be strings, or combinations of strings and numbers. For example, the first column, country_names,  certainly contains a bunch of strings.

If you want to know some summary statsitics about all numeric columns in this dataframe, you can use df.describe(). It will give you a dataframe that contains the summary statistics of the numbers.

In [ ]:
df.describe()

,country_id,year,project,historical,codingstart,codingend,codingstart_contemp,codingend_contemp,codingstart_hist,codingend_hist,...,e_mipopula,e_miurbani,e_pefeliex,e_wb_pop,e_pechmor,e_miinteco,e_civil_war,e_miinterc,e_pt_coup,e_pt_coup_attempts
count,27734.000000,27734.000000,27734.000000,27734.000000,27734.000000,27734.000000,26335.000000,26335.000000,16896.000000,16896.000000,...,1.856800e+04,17078.000000,7168.000000,1.018200e+04,24844.000000,21218.000000,13776.000000,21218.000000,10373.000000,10373.000000
mean,107.107305,1928.984784,0.409173,0.609216,1842.828009,2014.554518,1906.225973,2022.427910,1799.158736,1918.172112,...,1.849079e+04,0.256077,61.186155,3.261888e+07,235.310171,0.116269,0.061266,0.086483,0.023233,0.046756
std,82.120627,64.706575,0.591003,0.487935,58.323905,34.545471,16.915922,5.197873,19.870692,19.221587,...,7.458166e+04,0.225348,12.685666,1.193799e+08,165.911716,0.320555,0.239826,0.281083,0.160565,0.244556
min,3.000000,1789.000000,0.000000,0.000000,1789.000000,1859.000000,1900.000000,1948.000000,1789.000000,1859.000000,...,1.789700e+01,0.002000,30.200000,4.170000e+04,1.470000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,46.000000,1881.000000,0.000000,0.000000,1789.000000,2023.000000,1900.000000,2023.000000,1789.000000,1920.000000,...,1.032283e+03,0.119000,50.127500,2.367802e+06,64.900000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,91.000000,1939.000000,0.000000,1.000000,1815.000000,2023.000000,1900.000000,2023.000000,1789.000000,1920.000000,...,3.416651e+03,0.217000,62.371000,7.179500e+06,239.000000,0.000000,0.000000,0.000000,0.000000,0.000000
75%,155.000000,1984.000000,1.000000,1.000000,1900.000000,2023.000000,1900.000000,2023.000000,1800.000000,1920.000000,...,9.400638e+03,0.343000,72.646000,2.129082e+07,394.000000,0.000000,0.000000,0.000000,0.000000,0.000000
max,373.000000,2023.000000,2.000000,1.000000,2011.000000,2023.000000,2011.000000,2023.000000,1878.000000,2010.000000,...,1.262645e+06,2.590000,84.580000,1.417173e+09,756.000000,1.000000,1.000000,1.000000,3.000000,4.000000


That's all for data importing. Let's move on and talk about data exporting. If all you want to do is just to get some summary statistics, we might have reached the step to output the data so that we can present or communicate with others, although in real life, we might want to do more. To output data, see below.

In [ ]:
# First we store that summary statistics into a Python object.
sum_describe = df.describe()

**Again, these directories are mine. They are here just to make the codes work for me. Yours can be different.**

In [ ]:
# Then we output it to csv.
sum_describe.to_csv("/content/drive/MyDrive/POLI3148/POLI3148_Lecture4/data_processed/sum_describe.csv")

In [ ]:
# Alternatively, you can also output it as Excel file.
# You will need some more library to do it. I will comment it out for now.
# sum_describe.to_excel()

In [ ]:
# If you are interested, you can also save it as JSON file.
sum_describe.to_json("/content/drive/MyDrive/POLI3148/POLI3148_Lecture4/data_processed/sum_describe.json")

In [ ]:
# There is also a function to import JSON to Python
sum_describe = pd.read_json("/content/drive/MyDrive/POLI3148/POLI3148_Lecture4/data_processed/sum_describe.json")

As you have your first processed data, here are some reminders re good practice:

1. Separately store raw and processed data
2. NEVER overwrite raw data with processed data
3. Assign informative names to files of processed data, preferably similar to the objects’ names in your Python program
4. Saving intermediate outputs is good


To summarize:

`pd.read_csv()` loads data from designated path to Python.

`df.head()` gives you a dataframe that contains all columns and the first 5 rows of "df".

`df.tail()` gives you a dataframe that contains all columns and the last 5 rows of "df".

`df.info()` outputs info about all columns in this dataframe, including name and data type.

`df.describe()` gives you a dataframe that conatins some summary statistics of all numeric columns of "df".

`df.columns` reports columns of this df.

`df.shape` gives you a tuple of the number of rows and columns.

`df.to_csv()` saves the dataframe to a csv file at the specified path.

I know what you are thinking: Why sometimes there are parentheses and sometimes there isn't? Good question! df.shape gives us the rows and columns of df with a tuple, and df.columns gives us the column names. **They are attributes of this dataframe. We are not applying any method to this df when using .shape, so there is NO parantheses.** In contrast, df.head(), df.tail(), df.info(), df.describe() are all methods by definition, so we do include the parantheses.

## 3. Data Processing

While we are still focusing on the VDem data, I want to do something about it. I want to look at China's economic development, specifically GDP and GDP per Capita. Since we already imported the VDem, now what we need is columns that are relevant to our research question.

This is when having documentations and data source is super important. You can look at the readme file, or the data website to know what these columns mean. I did look at the documentation. GDP is coded e_gdp, and GDP per Capita is coded e_gdppc.

**Data subsetting**

So, the first thing we need to do, is to subset the dataframe to include columns and rows that are revelant. This is the time when we introduce data subsetting. If we want look at observations of China's economy from 1949 up until 2020, we will need to subset 1) observations of China and 2) observations from 1949 to 2020.

This gives you all rows about China:

In [ ]:
df.loc[df.country_name == "China"]

,country_name,country_text_id,country_id,year,historical_date,project,historical,histname,codingstart,codingend,...,e_mipopula,e_miurbani,e_pefeliex,e_wb_pop,e_pechmor,e_miinteco,e_civil_war,e_miinterc,e_pt_coup,e_pt_coup_attempts
16314,China,CHN,110,1789,1789-12-31,1,1,Great Qing Empire,1789,2023,...,NaN,NaN,NaN,NaN,NaN,1.0,NaN,1.0,NaN,NaN
16315,China,CHN,110,1790,1790-12-31,1,1,Great Qing Empire,1789,2023,...,NaN,NaN,NaN,NaN,NaN,1.0,NaN,0.0,NaN,NaN
16316,China,CHN,110,1791,1791-12-31,1,1,Great Qing Empire,1789,2023,...,NaN,NaN,NaN,NaN,NaN,1.0,NaN,0.0,NaN,NaN
16317,China,CHN,110,1792,1792-12-31,1,1,Great Qing Empire,1789,2023,...,NaN,NaN,NaN,NaN,NaN,0.0,NaN,0.0,NaN,NaN
16318,China,CHN,110,1793,1793-12-31,1,1,Great Qing Empire,1789,2023,...,NaN,NaN,NaN,NaN,NaN,0.0,NaN,0.0,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
16544,China,CHN,110,2019,2019-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.407745e+09,8.34,NaN,NaN,NaN,0.0,0.0
16545,China,CHN,110,2020,2020-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.411100e+09,8.09,NaN,NaN,NaN,0.0,0.0
16546,China,CHN,110,2021,2021-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.412360e+09,7.85,NaN,NaN,NaN,NaN,NaN
16547,China,CHN,110,2022,2022-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.412175e+09,7.61,NaN,NaN,NaN,NaN,NaN


This gives you all rows between year 1949 and 2020:

In [ ]:
df.loc[(df.year >= 1949) & (df.year<=2020)]

,country_name,country_text_id,country_id,year,historical_date,project,historical,histname,codingstart,codingend,...,e_mipopula,e_miurbani,e_pefeliex,e_wb_pop,e_pechmor,e_miinteco,e_civil_war,e_miinterc,e_pt_coup,e_pt_coup_attempts
160,Mexico,MEX,3,1949,1949-12-31,0,1,United Mexican States,1789,2023,...,27675.962,0.364,NaN,NaN,233.0,0.0,0.0,0.0,NaN,NaN
161,Mexico,MEX,3,1950,1950-12-31,0,1,United Mexican States,1789,2023,...,28485.180,0.359,NaN,NaN,224.0,0.0,0.0,0.0,0.0,0.0
162,Mexico,MEX,3,1951,1951-12-31,0,1,United Mexican States,1789,2023,...,29494.512,0.377,NaN,NaN,255.0,0.0,0.0,0.0,0.0,0.0
163,Mexico,MEX,3,1952,1952-12-31,0,1,United Mexican States,1789,2023,...,30503.844,0.394,NaN,NaN,239.0,0.0,0.0,0.0,0.0,0.0
164,Mexico,MEX,3,1953,1953-12-31,0,1,United Mexican States,1789,2023,...,31513.176,0.409,NaN,NaN,224.0,0.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
26327,Zanzibar,ZZB,236,2016,2016-12-31,0,1,"Zanzibar, semi-autonomous region of Tanzania",1856,2023,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
26328,Zanzibar,ZZB,236,2017,2017-12-31,0,1,"Zanzibar, semi-autonomous region of Tanzania",1856,2023,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
26329,Zanzibar,ZZB,236,2018,2018-12-31,0,1,"Zanzibar, semi-autonomous region of Tanzania",1856,2023,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
26330,Zanzibar,ZZB,236,2019,2019-12-31,0,1,"Zanzibar, semi-autonomous region of Tanzania",1856,2023,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


And this gives you the intersection of both:

In [ ]:
df.loc[(df.country_name == "China") & (df.year >= 2000) & (df.year <= 2020)]

,country_name,country_text_id,country_id,year,historical_date,project,historical,histname,codingstart,codingend,...,e_mipopula,e_miurbani,e_pefeliex,e_wb_pop,e_pechmor,e_miinteco,e_civil_war,e_miinterc,e_pt_coup,e_pt_coup_attempts
16525,China,CHN,110,2000,2000-12-31,0,1,People's Republic of China,1789,2023,...,1262645.0,NaN,72.93,1.262645e+09,36.80,0.0,0.0,0.0,0.0,0.0
16526,China,CHN,110,2001,2001-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.271850e+09,34.20,NaN,0.0,NaN,0.0,0.0
16527,China,CHN,110,2002,2002-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.280400e+09,31.50,NaN,0.0,NaN,0.0,0.0
16528,China,CHN,110,2003,2003-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.288400e+09,28.90,NaN,0.0,NaN,0.0,0.0
16529,China,CHN,110,2004,2004-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.296075e+09,26.30,NaN,0.0,NaN,0.0,0.0
16530,China,CHN,110,2005,2005-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.303720e+09,24.00,NaN,0.0,NaN,0.0,0.0
16531,China,CHN,110,2006,2006-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.311020e+09,21.90,NaN,0.0,NaN,0.0,0.0
16532,China,CHN,110,2007,2007-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.317885e+09,20.10,NaN,NaN,NaN,0.0,0.0
16533,China,CHN,110,2008,2008-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.324655e+09,18.50,NaN,NaN,NaN,0.0,0.0
16534,China,CHN,110,2009,2009-12-31,0,1,People's Republic of China,1789,2023,...,NaN,NaN,NaN,1.331260e+09,17.00,NaN,NaN,NaN,0.0,0.0


NOTE: You CANNOT use the base Python keywords such as "and", "or". They won't work. You need to use logic denotations, such as &, and |.

Now, we have successfully subsetted the rows that we want. But remember, there are over 4 thousands columns. We certainly do not want to deal with that many columns. How do we subset columns? The answer is also df.loc. Let's look at subsetting columns only first.

In [ ]:
df.loc[:,"country_name"]

,country_name
0,Mexico
1,Mexico
2,Mexico
3,Mexico
4,Mexico
...,...
27729,Piedmont-Sardinia
27730,Piedmont-Sardinia
27731,Piedmont-Sardinia
27732,Piedmont-Sardinia


You just got one column. If you want multiple columns, you need to pass in a list of strings of column names.

In [ ]:
df.loc[:,["country_name", "year", "e_gdp", "e_gdppc"]]

,country_name,year,e_gdp,e_gdppc
0,Mexico,1789,871.938,1.352
1,Mexico,1790,878.908,1.357
2,Mexico,1791,888.411,1.371
3,Mexico,1792,902.880,1.389
4,Mexico,1793,912.872,1.401
...,...,...,...,...
27729,Piedmont-Sardinia,1857,NaN,NaN
27730,Piedmont-Sardinia,1858,NaN,NaN
27731,Piedmont-Sardinia,1859,NaN,NaN
27732,Piedmont-Sardinia,1860,NaN,NaN


The syntax of .loc works as follows:

* df.loc[row-wise subsetting criteria, column-wise subsetting criteria]

* “:” means SELECT ALL

* df.loc[:, ['country_name', 'year', 'e_gdp', 'e_gdppc']] means, “select all rows, and subset this list of columns”

Q: Wait, but when we subset row (e.g., df.loc[df.country_name == "China"]), we used no “:”.

A: The above expression is an abbreviated version of df.loc[df.country_name == "China", :].

Pandas allow us to omit the subsetting conditions regarding columns if subset by row only.

However, if we subset by column only, we cannot omit the subsetting condition regarding rows.

Q: What are the inputs of the selection criteria?

A: Column/ row names, put in a list. E.g., ['country_name', 'year', 'e_gdp', 'e_gdppc']

Or, A list/ array/ column of booleans (Trae/ False). True means the row/ column should be included, False means the the row/ column should be excluded.
Note: It should have the same length as the dimension

Now, let's piece everything together to subset the dataframe and get:

1. Observations that are related to China
2. Observations that are between 1949 and 2020
3. Variables that are about economy, i.e., e_gdp and e_gdppc

Observation is about rows, so the first two criteria go in the first slot. Variables are about columns, so the last one goes into the second slot.

In [ ]:
df.loc[(df.year<=2020) & (df.year>=1949) & (df.country_name == "China"), ["country_name", "year", "e_gdp", "e_gdppc"]]

,country_name,year,e_gdp,e_gdppc
16474,China,1949,55706.297,0.956
16475,China,1950,55198.256,0.936
16476,China,1951,58664.426,0.979
16477,China,1952,63147.690,1.036
16478,China,1953,68123.152,1.091
...,...,...,...,...
16541,China,2016,2039529.494,13.890
16542,China,2017,2136176.436,14.485
16543,China,2018,2205729.606,14.948
16544,China,2019,2279809.269,15.384


If you find it very hard to write and read, you can also create a manual filter before doing the subsetting. This is my personal favourite way to subset. Note you cannot use the builtin keyword "filter". That will override the builtin function.

In [ ]:
filt = (df.year<=2020) & (df.year>=1949) & (df.country_name == "China")
df.loc[filt, ["country_name", "year", "e_gdp", "e_gdppc"]]

,country_name,year,e_gdp,e_gdppc
16474,China,1949,55706.297,0.956
16475,China,1950,55198.256,0.936
16476,China,1951,58664.426,0.979
16477,China,1952,63147.690,1.036
16478,China,1953,68123.152,1.091
...,...,...,...,...
16541,China,2016,2039529.494,13.890
16542,China,2017,2136176.436,14.485
16543,China,2018,2205729.606,14.948
16544,China,2019,2279809.269,15.384


Alternatively, Prof. Chen also demonstrated another way, which is to break the long line of code into several lines using the slash.

In [ ]:
df_china_econ = df.loc[(df.country_name == "China") \
       & (df.year >= 1949) \
       & (df.year <= 2020) \
       , ["country_name", "year", "e_gdp", "e_gdppc"]]

Okay! That is all we need to know about subsetting. There are some good practices that Prof. Chen talked about.

1. Subset step by step. Don’t do all operations in one command.
2. Use comments to document the subsetting operation
3. Save the subset of interest for your data analysis
4. Avoid repeatedly clean and wrangle the data
5. Use informative file names
6. Keep the original/ raw data, even if you don’t think you need them any more.
E.g., if you conduct a study on China and do the example operations for subset, you are strongly recommended to keep the global data

**Data Transforming**

After we obtained this China Economy dataframe, we might want to transform this dataframe so that it is easier for us to perform some analysis.

First, we want to transform a numeric variable to another numeric variable. Let's say the GDP numbers are too large and we want to move the decimal point forward by three, i.e., dividing GDP by 1000. How do we do that?

We simply create another column using exisitng columns, like this:

In [ ]:
df_china_econ["gdp_1000"] = df_china_econ["e_gdp"]/1000

The above is the same as:

In [ ]:
df_china_econ["gdp_1000"] = df_china_econ.e_gdp/1000

That is to say, when you refer to an existing column inside a dataframe, you can use the attribute-like way. If you want to create a column, the canonical square bracket is the only way to go.

Next, sometimes economists work with log data to capture some pattern. What if we want to transform the GDP value to a log scale? To do that, we need another library called numpy.

In [ ]:
import numpy as np
df_china_econ['gdp_log'] = np.log(df_china_econ['e_gdp'])

Now let's look at what the dataframe is like:

In [ ]:
df_china_econ

,country_name,year,e_gdp,e_gdppc,gdp_1000,gdp_log
16474,China,1949,55706.297,0.956,55.706297,10.927848
16475,China,1950,55198.256,0.936,55.198256,10.918687
16476,China,1951,58664.426,0.979,58.664426,10.979589
16477,China,1952,63147.690,1.036,63.147690,11.053232
16478,China,1953,68123.152,1.091,68.123152,11.129072
...,...,...,...,...,...,...
16541,China,2016,2039529.494,13.890,2039.529494,14.528230
16542,China,2017,2136176.436,14.485,2136.176436,14.574528
16543,China,2018,2205729.606,14.948,2205.729606,14.606569
16544,China,2019,2279809.269,15.384,2279.809269,14.639602


Next! Now we want to transform a numeric variable to binary variable. We want to add another column that marks if the year is under Xi's administration.

In [ ]:
df_china_econ["admin_xi"] = (df_china_econ["year"] <= 2023) & (df_china_econ["year"] >= 2013)

In [ ]:
df_china_econ

,country_name,year,e_gdp,e_gdppc,gdp_1000,gdp_log,admin_xi
16474,China,1949,55706.297,0.956,55.706297,10.927848,False
16475,China,1950,55198.256,0.936,55.198256,10.918687,False
16476,China,1951,58664.426,0.979,58.664426,10.979589,False
16477,China,1952,63147.690,1.036,63.147690,11.053232,False
16478,China,1953,68123.152,1.091,68.123152,11.129072,False
...,...,...,...,...,...,...,...
16541,China,2016,2039529.494,13.890,2039.529494,14.528230,True
16542,China,2017,2136176.436,14.485,2136.176436,14.574528,True
16543,China,2018,2205729.606,14.948,2205.729606,14.606569,True
16544,China,2019,2279809.269,15.384,2279.809269,14.639602,True


That does not look pretty nice. I want to have all Chinese presidents in this column. I need to remove the admin_xi column first.

In [ ]:
df_china_econ.drop(columns = ['admin_xi'], inplace = True)

You can also do it using assignment:



```
df_china_econ = df_china_econ.drop(columns = ['admin_xi'])
```

And you can drop multiple columns at once, as long as you pass in a list for the columns argument.

```
df_china_econ = df_china_econ.drop(columns = ['gdp_1000', 'gdp_log', 'admin_xi'])
```




In [ ]:
df_china_econ

,country_name,year,e_gdp,e_gdppc,gdp_1000,gdp_log
16474,China,1949,55706.297,0.956,55.706297,10.927848
16475,China,1950,55198.256,0.936,55.198256,10.918687
16476,China,1951,58664.426,0.979,58.664426,10.979589
16477,China,1952,63147.690,1.036,63.147690,11.053232
16478,China,1953,68123.152,1.091,68.123152,11.129072
...,...,...,...,...,...,...
16541,China,2016,2039529.494,13.890,2039.529494,14.528230
16542,China,2017,2136176.436,14.485,2136.176436,14.574528
16543,China,2018,2205729.606,14.948,2205.729606,14.606569
16544,China,2019,2279809.269,15.384,2279.809269,14.639602


Next, I need to create a new column where each year has a name of the presidents, i.e., Mao, Hua-Hu-Zhao, Jiang, Hu, Xi.

* Mao: 1949-1976
* Hua-Hu-Zhao: 1976-1989
* Jiang: 1989-2002
* Hu: 2002-2013
* Xi: 2013-2023


In [ ]:
df_china_econ['admin'] = None # Initiate a column named admin with None type
df_china_econ.loc[(df_china_econ.year >= 1949) & (df_china_econ.year < 1976), 'admin'] = "Mao"
df_china_econ.loc[(df_china_econ.year >= 1976) & (df_china_econ.year < 1989), 'admin'] = "Hua-Hu-Zhao"
df_china_econ.loc[(df_china_econ.year >= 1989) & (df_china_econ.year < 2002), 'admin'] = "Jiang"
df_china_econ.loc[(df_china_econ.year >= 2002) & (df_china_econ.year < 2013), 'admin'] = "Hu"
df_china_econ.loc[(df_china_econ.year >= 2013) & (df_china_econ.year <= 2023), 'admin'] = "Xi"

In [ ]:
df_china_econ

,country_name,year,e_gdp,e_gdppc,gdp_1000,gdp_log,admin
16474,China,1949,55706.297,0.956,55.706297,10.927848,Mao
16475,China,1950,55198.256,0.936,55.198256,10.918687,Mao
16476,China,1951,58664.426,0.979,58.664426,10.979589,Mao
16477,China,1952,63147.690,1.036,63.147690,11.053232,Mao
16478,China,1953,68123.152,1.091,68.123152,11.129072,Mao
...,...,...,...,...,...,...,...
16541,China,2016,2039529.494,13.890,2039.529494,14.528230,Xi
16542,China,2017,2136176.436,14.485,2136.176436,14.574528,Xi
16543,China,2018,2205729.606,14.948,2205.729606,14.606569,Xi
16544,China,2019,2279809.269,15.384,2279.809269,14.639602,Xi


**Data Aggregating**

Now everything looks pretty good. I need some statistics to see which years are the best and which ones are the worst in terms of economic performance. So it is a good time for us to introduce "aggregate".

If you are familiar with R, this notion could be very straightforward to you. Aggregation in Python is pretty much like summarise() in R, which more limited functionalities.

In [ ]:
df_china_econ["gdp_1000"].agg(['mean', 'std', 'min', 'max'])

,gdp_1000
mean,540.911609
std,635.191848
min,55.198256
max,2279.809269


If you want to group by administration, you can add in this groupby method. It will give you these statistics for each value of the column grouped. So here, for each value inside the admin column, you get mean, std, min, and max.



In [ ]:
df_china_econ.groupby('admin')['gdp_1000'].agg(['mean', 'std', 'min', 'max'])

,mean,std,min,max
admin,,,,
Hu,1165.359951,313.400551,738.392701,1662.715637
Hua-Hu-Zhao,220.689279,59.793327,144.157420,322.595831
Jiang,471.603783,123.493054,315.683097,682.140585
Mao,87.463739,25.922919,55.198256,142.770886
Xi,2032.062001,189.792406,1758.258459,2279.809269


What if you want different statistics for different columns? Can you have different aggragation for each of the columns? Of course. See below:

In [ ]:
df_china_econ.agg({'gdp_1000': ['mean', 'min', 'max'], 'e_gdppc': ['mean', 'min', 'max', 'std']})

gdp_1000                              e_gdppc                  \
                    mean          min          max       mean     min     max   
admin                                                                           
Hu           1165.359951   738.392701  1662.715637   8.272364   5.427  11.527   
Hua-Hu-Zhao   220.689279   144.157420   322.595831   2.023462   1.462   2.733   
Jiang         471.603783   315.683097   682.140585   3.662692   2.626   5.051   
Mao            87.463739    55.198256   142.770886   1.147778   0.936   1.464   
Xi           2032.062001  1758.258459  2279.809269  13.843571  12.129  15.384   

                       
                  std  
admin                  
Hu           2.076375  
Hua-Hu-Zhao  0.431904  
Jiang        0.817237  
Mao          0.155961  
Xi           1.186162

In [ ]:
df_china_econ.groupby('admin').agg({'gdp_1000': ['mean', 'min', 'max'], 'e_gdppc': ['mean', 'min', 'max', 'std']})

Dr Chen recommends this approach. It looks nicer.

In [ ]:
df_china_econ.groupby('admin').agg(
    gdp_mean = ('gdp_1000', 'mean'),
    gdp_min = ('gdp_1000', 'min'),
    gdp_max = ('gdp_1000', 'max'),
    gdp_std = ('e_gdppc', 'std'),
    gdppc_mean = ('e_gdppc', 'mean'),
    gdppc_min = ('e_gdppc', 'min'),
    gdppc_max = ('e_gdppc', 'max'),
    gdppc_sd = ('e_gdppc', 'std')
).reset_index()

,admin,gdp_mean,gdp_min,gdp_max,gdp_std,gdppc_mean,gdppc_min,gdppc_max,gdppc_sd
0,Hu,1165.359951,738.392701,1662.715637,2.076375,8.272364,5.427,11.527,2.076375
1,Hua-Hu-Zhao,220.689279,144.157420,322.595831,0.431904,2.023462,1.462,2.733,0.431904
2,Jiang,471.603783,315.683097,682.140585,0.817237,3.662692,2.626,5.051,0.817237
3,Mao,87.463739,55.198256,142.770886,0.155961,1.147778,0.936,1.464,0.155961
4,Xi,2032.062001,1758.258459,2279.809269,1.186162,13.843571,12.129,15.384,1.186162


The reason why we reset the index is that we do not want those values inside the admin column to be the index. Python automatically set the row index to be the values that are grouped. By reset_index, we create another set of integers from 0 to be our index for each row.

In [ ]:
sum_by_admin = df_china_econ.groupby('admin').agg(
    gdp_mean = ('gdp_1000', 'mean'),
    gdp_min = ('gdp_1000', 'min'),
    gdp_max = ('gdp_1000', 'max'),
    gdp_std = ('gdp_1000', 'std'),

    gdppc_mean = ('e_gdppc', 'mean'),
    gdppc_min = ('e_gdppc', 'min'),
    gdppc_max = ('e_gdppc', 'max'),
    gdppc_std = ('e_gdppc', 'std'),
).reset_index()

In [ ]:
sum_by_admin

Let's save that as sum_by_admin. Another function we want to let you know is idxmax() and idxmin(). They give you the index of first occurrence of maximum over requested axis. NA/null values are excluded. For example, if you want to know when China performed the worst or the best under the administration of each leaders, you can do the following:

In [ ]:
df_china_econ.loc[df_china_econ.groupby('admin').gdp_1000.idxmin()]

df_china_econ.groupby('admin').gdp_1000.idxmin()

In [ ]:
df_china_econ.groupby('admin').gdp_1000.idxmax()

**Data Renaming**

New task: We want to compare the economy of China and Japan from 1949 to 2019.

What we need: Create a new subset named df_japan_econ. In the same way, we filter out the country name of Japan, with year of 1949 to 2019, and only include columns that are relevant.

In [ ]:
df_japan_econ = df.loc[(df.country_name == "Japan") & (df.year >= 1949) & \
                        (df.year <= 2019), \
                         ['country_name', 'year', 'e_gdp', 'e_gdppc']]

In [ ]:
df_japan_econ

Now because we have two different countries, we might need to rename their column names so that there is no confusion when the data is joined together. How do you rename variables? Use dictionary! **You make the original column name the key of that dictionary, and the desired new name the value of that dictionary.** And use the rename method.

In [ ]:
{'e_gdp': 'gdp_china', 'e_gdppc': 'gdppc_china'}

In [ ]:
df_china_econ_h = df_china_econ[['year', 'e_gdp', 'e_gdppc']].\
  rename(columns = {'e_gdp': 'gdp_china', 'e_gdppc': 'gdppc_china'})

In [ ]:
df_china_econ_h

In [ ]:
df_japan_econ_h = df_china_econ[['year', 'e_gdp', 'e_gdppc']].\
  rename(columns = {'e_gdp': 'gdp_japan', 'e_gdppc': 'gdppc_japan'})

In [ ]:
df_japan_econ_h

**Data Combining**

To combine datasets horizontally (i.e., placing columns of two datasets side by side), we use pd.merge().

In [ ]:
df_cn_jp = pd.merge(df_china_econ_h, df_japan_econ_h, on = 'year')

By contrast, to combine datasets vertically (i.e., placing data entries of dataset B at the bottom of dataset A), we use pd.concat().

In [ ]:
df_china_econ_v = df_china_econ.loc[:, ['country_name', 'year', 'e_gdp', 'e_gdppc']]
df_japan_econ_v = df_japan_econ.loc[:, ['country_name', 'year', 'e_gdp', 'e_gdppc']]

In [ ]:
pd.concat([df_china_econ_v, df_japan_econ_v])

## 4. Logistics

Continue working on the in-class exercise 3. It will be due next Monday at 23:59.